servers/rendering/renderer_rd/framebuffer_cache_rd.h

```diff
@@ -0,0 +1,310 @@
+		return _hash_rids(h, args...);
+	}
+
+	_FORCE_INLINE_ bool _compare_args(uint32_t idx, const LocalVector<RID> &textures, const RID &arg) {
+		return textures[idx] == arg;
+	}
+
+	template <typename... Args>
+	_FORCE_INLINE_ bool _compare_args(uint32_t idx, const LocalVector<RID> &textures, const RID &arg, Args... args) {
+		if (textures[idx] != arg) {
+			return false;
+		}
+		return _compare_args(idx + 1, textures, args...);
+	}
+
+	_FORCE_INLINE_ void _create_args(Vector<RID> &textures, const RID &arg) {
+		textures.push_back(arg);
+	}
+
+	template <typename... Args>
+	_FORCE_INLINE_ void _create_args(Vector<RID> &textures, const RID &arg, Args... args) {
+		textures.push_back(arg);
+		_create_args(textures, args...);
+	}
+
+	static FramebufferCacheRD *singleton;
+
+	uint32_t cache_instances_used = 0;
+
+	void _invalidate(Cache *p_cache);
+	static void _framebuffer_invalidation_callback(void *p_userdata);
+
+	RID _allocate_from_data(uint32_t p_views, uint32_t p_hash, uint32_t p_table_idx, const Vector<RID> &p_textures, const Vector<RD::FramebufferPass> &p_passes) {
+		RID rid;
+		if (p_passes.size()) {
+			rid = RD::get_singleton()->framebuffer_create_multipass(p_textures, p_passes, RD::INVALID_ID, p_views);
+		} else {
+			rid = RD::get_singleton()->framebuffer_create(p_textures, RD::INVALID_ID, p_views);
+		}
+
+		ERR_FAIL_COND_V(rid.is_null(), rid);
+
+		Cache *c = cache_allocator.alloc();
+		c->views = p_views;
+		c->cache = rid;
+		c->hash = p_hash;
+		c->textures.resize(p_textures.size());
+		for (uint32_t i = 0; i < c->textures.size(); i++) {
+			c->textures[i] = p_textures[i];
+		}
+		c->passes.resize(p_passes.size());
+		for (uint32_t i = 0; i < c->passes.size(); i++) {
+			c->passes[i] = p_passes[i];
+		}
+		c->prev = nullptr;
+		c->next = hash_table[p_table_idx];
+		if (hash_table[p_table_idx]) {
+			hash_table[p_table_idx]->prev = c;
+		}
+		hash_table[p_table_idx] = c;
+
+		RD::get_singleton()->framebuffer_set_invalidation_callback(rid, _framebuffer_invalidation_callback, c);
+
+		cache_instances_used++;
+
+		return rid;
+	}
+
+public:
+	template <typename... Args>
+	RID get_cache(Args... args) {
+		uint32_t h = hash_murmur3_one_32(1); //1 view
+		h = hash_murmur3_one_32(sizeof...(Args), h);
+		h = _hash_args(h, args...);
+		h = hash_murmur3_one_32(0, h); // 0 passes
+		h = hash_fmix32(h);
+
+		uint32_t table_idx = h % HASH_TABLE_SIZE;
+		{
+			const Cache *c = hash_table[table_idx];
+
+			while (c) {
+				if (c->hash == h && c->passes.size() == 0 && c->textures.size() == sizeof...(Args) && c->views == 1 && _compare_args(0, c->textures, args...)) {
+					return c->cache;
+				}
+				c = c->next;
+			}
+		}
+
+		// Not in cache, create:
+
+		Vector<RID> textures;
+		_create_args(textures, args...);
+
+		return _allocate_from_data(1, h, table_idx, textures, Vector<RD::FramebufferPass>());
+	}
+
+	template <typename... Args>
+	RID get_cache_multiview(uint32_t p_views, Args... args) {
+		uint32_t h = hash_murmur3_one_32(p_views);
```

Comment:
Not important, but an `ERR_FAIL_COND_V(p_views >= 2, RID())` may help catch some errors?

Reply:
I think you can technically have as many views as you want, if you restrict to 2 then its more like an implementation dependant thing.